cirq/google/api/v2/results.py

```diff
@@ -0,0 +1,184 @@
+            hardware so that we can combine measurements that occupy the same
+            slot.
+    """
+
+
+def find_measurements(program: Union[circuits.Circuit, schedules.Schedule],
+                     ) -> List[MeasureInfo]:
+    """Find measurements in the given program (circuit or schedule).
+
+    Returns:
+        List of Measurement objects with named measurements in this program.
+    """
+    measurements: List[MeasureInfo] = []
+    keys: Set[str] = set()
+
+    if isinstance(program, circuits.Circuit):
+        measure_iter = _circuit_measurements(program)
+    else:
+        measure_iter = _schedule_measurements(program)
+
+    for m in measure_iter:
+        if m.key in keys:
+            raise ValueError('Duplicate measurement key: {}'.format(m.key))
+        keys.add(m.key)
+        measurements.append(m)
+
+    return measurements
+
+
+def _circuit_measurements(circuit: circuits.Circuit) -> Iterator[MeasureInfo]:
+    for i, moment in enumerate(circuit):
+        for op in moment:
+            if (isinstance(op, ops.GateOperation) and
+                    isinstance(op.gate, ops.MeasurementGate)):
+                yield MeasureInfo(key=op.gate.key,
+                                  qubits=_grid_qubits(op),
+                                  slot=i)
+
+
+def _schedule_measurements(schedule: schedules.Schedule
+                          ) -> Iterator[MeasureInfo]:
+    for so in schedule.scheduled_operations:
+        op = so.operation
+        if (isinstance(op, ops.GateOperation) and
+                isinstance(op.gate, ops.MeasurementGate)):
+            yield MeasureInfo(key=op.gate.key,
+                              qubits=_grid_qubits(op),
+                              slot=so.time.raw_picos())
+
+
+def _grid_qubits(op: ops.Operation) -> List[devices.GridQubit]:
+    if not all(isinstance(q, devices.GridQubit) for q in op.qubits):
+        raise ValueError('Expected GridQubits: {}'.format(op.qubits))
+    return cast(List[devices.GridQubit], list(op.qubits))
+
+
+def pack_bits(bits: np.ndarray) -> bytes:
+    """Pack bits given as a numpy array of bools into bytes."""
+    # Pad length to multiple of 8 if needed.
+    remainder = len(bits) % 8
+    if remainder:
+        bits = np.pad(bits, (0, 8 - remainder), 'constant')
+
+    # Pack in little-endian bit order.
+    bits = bits.reshape((-1, 8))[:, ::-1]
+    byte_arr = np.packbits(bits, axis=1).reshape(-1)
+
+    return byte_arr.tobytes()
+
+
+def unpack_bits(data: bytes, repetitions: int) -> np.ndarray:
+    """Unpack bits from a byte array into numpy array of bools."""
+    byte_arr = np.frombuffer(data, dtype='uint8').reshape((len(data), 1))
+    bits = np.unpackbits(byte_arr, axis=1)[:, ::-1].reshape(-1).astype(bool)
+    return bits[:repetitions]
+
+
+def results_to_proto(
+        trial_sweeps: Iterable[Iterable[study.TrialResult]],
+        measurements: List[MeasureInfo],
+        msg: Optional[result_pb2.Result] = None,
```

Comment:
Nit: I would rename this to `out` and force keyword args on it, which is how numpy does it.

Reply:
Good idea. Done.